File: internal/app/repo/google/google.go
package google

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"

	"go.uber.org/dig"
	"golang.org/x/oauth2"
)

type (
	GoogleCallbackRequest struct {
		State string `json:"state"`
		Code  string `json:"code"`
	}

	GoogleCallbackResponse struct {
		ID            string `json:"id"`
		Email         string `json:"email"`
		Name          string `json:"name"`
		Picture       string `json:"picture"`
		VerifiedEmail bool   `json:"verified_email"`
	}

	GoogleRepo interface {
		Login(ctx context.Context) (url string, err error)
		Callback(ctx context.Context, req GoogleCallbackRequest) (dataInfo GoogleCallbackResponse, err error)
	}

	GoogleRepoImpl struct {
		dig.In

		*oauth2.Config
	}
)

func NewGoogleRepo(impl GoogleRepoImpl) GoogleRepo {
	return &impl
}

func (g *GoogleRepoImpl) Login(ctx context.Context) (url string, err error) {
	url = g.AuthCodeURL("state-token", oauth2.AccessTypeOffline)
	return
}

func (g *GoogleRepoImpl) Callback(ctx context.Context, req GoogleCallbackRequest) (dataInfo GoogleCallbackResponse, err error) {
	if req.State != "state-token" {
		err = fmt.Errorf("invalid credentials")
		slog.ErrorContext(ctx, "[repo][google][Callback][state]", err)
		return
	}

	token, err := g.Exchange(ctx, req.Code)
	if err != nil {
		slog.ErrorContext(ctx, "[repo][google][Callback][Exchange]", err)
		return
	}

	resp, err := http.Get("https://www.googleapis.com/oauth2/v2/userinfo?access_token=" + token.AccessToken)
	if err != nil {
		slog.ErrorContext(ctx, "[repo][google][Callback][http.Get]", err)
		return
	}
	defer resp.Body.Close()

	userData, err := io.ReadAll(resp.Body)
	if err != nil {
		slog.ErrorContext(ctx, "[repo][google][Callback][io.ReadAll]", err)
		return
	}
	err = json.Unmarshal(userData, &dataInfo)
	if err != nil {
		slog.ErrorContext(ctx, "[repo][google][Callback][json.Unmarshal]", err)
		return
	}

	return
}

File: src/repo/google/google.ts
// src/repos/GoogleRepo.ts
import { OAuth2Client } from "google-auth-library";
import axios from "axios";

export interface GoogleCallbackRequest {
  state: string;
  code: string;
}

export interface GoogleCallbackResponse {
  id: string;
  email: string;
  name: string;
  picture: string;
  verified_email: boolean;
}

export class GoogleRepository {
  private oauth2Client: OAuth2Client;

  constructor() {
    this.oauth2Client = new OAuth2Client({
      clientId: process.env.GOOGLE_CLIENT_ID!,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET!,
      redirectUri: process.env.GOOGLE_REDIRECT_URI!,
    });
  }

  // --- LOGIN (Generate OAuth2 URL) ---
  public async login(): Promise<string> {
    const url = this.oauth2Client.generateAuthUrl({
      access_type: "offline",
      scope: [
        "https://www.googleapis.com/auth/userinfo.email",
        "https://www.googleapis.com/auth/userinfo.profile",
      ],
      state: "state-token",
    });
    return url;
  }

  // --- CALLBACK (Exchange Code & Get User Info) ---
  public async callback(
    req: GoogleCallbackRequest
  ): Promise<GoogleCallbackResponse> {
    if (req.state !== "state-token") {
      throw new Error("invalid credentials");
    }

    // Tukar auth code dengan access token
    const { tokens } = await this.oauth2Client.getToken(req.code);
    this.oauth2Client.setCredentials(tokens);

    if (!tokens.access_token) {
      throw new Error("No access token received from Google");
    }

    // Panggil Google userinfo endpoint
    const resp = await axios.get(
      "https://www.googleapis.com/oauth2/v2/userinfo",
      {
        headers: { Authorization: `Bearer ${tokens.access_token}` },
      }
    );

    return resp.data as GoogleCallbackResponse;
  }
}
